JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -10,91 +10,307 @@
+		Map<String, String> formData = new HashMap<String, String>();
+		formData.put("grant_type", "password");
+		formData.put("client_id", prop.getProperty("ClientId"));
+		formData.put("client_secret", prop.getProperty("ClientSecret"));
+		formData.put("username", prop.getProperty("User"));
+		formData.put("password", prop.getProperty("Password"));
+		GetAndSetTokens(formData);
+	}
+
+	// refresh grant uses the refresh token to get the new access and refresh token
+	public static void GetTokensByRefreshGrant() throws Exception {
+		Map<String, String> formData = new HashMap<String, String>();
+		formData.put("grant_type", "refresh_token");
+		formData.put("client_id", prop.getProperty("ClientId"));
+		formData.put("client_secret", prop.getProperty("ClientSecret"));
+		formData.put("refresh_token", System.getProperty(OAUTH_REFRESH_TOKEN_VAR_STRING));
+		GetAndSetTokens(formData);
+	}
+
+	// read result from Buffered input stream
+	private static ByteArrayOutputStream readResult(BufferedInputStream in, ByteArrayOutputStream buf) {
+		try {
+			for (int result = in.read(); result != -1; result = in.read()) {
+				buf.write((byte) result);
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+		return buf;
+	}
+
+	// get and set tokens from IDP
+	private static void GetAndSetTokens(Map<String, String> formData) throws Exception {
+		try {
+			String postOpts = getEncodedParamsString(formData);
+			byte[] postData = postOpts.getBytes("UTF-8");
+			int postDataLength = postData.length;
+			URL url = new URL(prop.getProperty("TokenUrl"));
+			HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+			try {
+				connection.setDoOutput(true);
+				connection.setUseCaches(false);
+				connection.setRequestMethod("POST");
+				connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+				connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+				connection.setRequestProperty("Accept", "application/json");
+				connection.getOutputStream().write(postData);
+				BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
+				ByteArrayOutputStream buf = new ByteArrayOutputStream();
+				readResult(in, buf);
+				String res = buf.toString("UTF-8");
+				JsonElement jElement = JsonParser.parseString(res);
+				JsonObject jObject = jElement.getAsJsonObject();
+				// set Tokens as System Properties - new values to access_token and
+				// refresh_token
+				String accessToken = jObject.has("access_token") ? jObject.get("access_token").getAsString() : null;
+				String refreshToken = jObject.has("refresh_token") ? jObject.get("refresh_token").getAsString() : null;
+				if (null == accessToken) {
+					throw new Exception(
+							"Access/refresh token is null, Please verify the config.properties for proper inputs.");
+				}
+				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, accessToken);
+				System.setProperty(OAUTH_REFRESH_TOKEN_VAR_STRING, refreshToken);
```

Comment:
What happens if the refresh token is null? Maybe we need a null check here.

Reply:
It will assign null or empty string to refreshToken.

Reply:
I think storing an environment variable as 'null' is a bad idea.

Reply:
**Added the null and empty check condition to check the refresh token, and store only if its not null.**